libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java

```diff
@@ -114,13 +116,19 @@ public XContent xContent() {
         }
     };
 
+    private static final Pattern COMPATIBLE_API_HEADER_PATTERN = Pattern.compile(
+        "(application|text)/(vnd.elasticsearch\\+)?([^;]+)(\\s*;\\s*compatible-with=(\\d+))?",
```

Comment:
Regex needs a few more escapes:
- Escape the `.` within `vnd.elasticsearch+`
- Escape the `/` between the type and subtype

 `(application|text)\\/(vnd\\.elasticsearch\\+)?([^;]+)(\\s*;\\s*compatible-with=(\\d+))?`

Reply:
Something I am curious about is the `(\\d+))?` at the end. If possible, it would be nice to define this as part of our "spec". This would be something we could be strict (eschew leniency!) about, such as enforcing it to be in a specific format (and throwing an exception when someone specifies a pattern that is not correct.

If you agree with the sentiment, I would be curious what we expect the final number to be, for example, which of these do we intend to support (now *and* in the future):

- compatible-with=7
- compatible-with=7.11
- compatible-with=7.x
- compatible-with=7.11.2

Then I think we should make this strict so users find out early if they've send an illegal value

Reply:
@sethmlarson Why does the forward slash needs escaping? It's not special in Java?

I wondering if we should anchor the pattern with `^` at the start and `$` at the end, to be on the safe side?

In `([^;]+)(\\s*;`, the `[^;]+` will also consume whitespace, so I believe you can remove the `\\s*`.

Can we switch any of these capture groups to non-capturing if we don't actually need the contents, just the grouping? i.e `(?:<pattern>)`

Reply:
@pugnascotia You're right the `/` doesn't need an escape in Java